R/get_draws.R

```diff
@@ -0,0 +1,444 @@
+#' draws. It could be any from `"all"` `"predicted_counts"`, `"predicted_ww"`,
+#' `"global_rt"`, or `"subpop_rt"`. When `what = "all"` (the default),
+#' the function will extract all four variables.
+#' @return  A tibble containing the full set of posterior draws of the
+#' estimated, nowcasted, and forecasted: counts, site-level wastewater
+#' concentrations, "global"(e.g. state) R(t) estimate, and the  "local" (site +
+#' the one auxiliary subpopulation) R(t) estimates. In the instance where there
+#' are observations, the data will be joined to each draw of the predicted
+#' observation to facilitate plotting.
+#' @export
+get_draws <- function(x, ..., what = "all") {
+  UseMethod("get_draws")
+}
+
+#' @rdname get_draws
+#' @details
+#' The function `get_draws_df()` has been deprecated in favor of `get_draws()`.
+#'
+#' @export
+get_draws_df <- function(x, ...) {
+  .Deprecated("get_draws")
+}
+
+#' S3 method for extracting posterior draws alongside data for a
+#' wwinference_fit object
+#'
+#' This method overloads the generic `get_draws` function specifically
+#' for objects of type 'wwinference_fit'.
+#'
+#' @rdname get_draws
+#' @export
+get_draws.wwinference_fit <- function(x, ..., what = "all") {
+  get_draws.data.frame(
+    x = x$raw_input_data$input_ww_data,
+    count_data = x$raw_input_data$input_count_data,
+    stan_data_list = x$stan_data_list,
+    fit_obj = x$fit,
+    what = what
+  )
+}
+
+#' @export
+#' @rdname get_draws
+get_draws.default <- function(x, ..., what = "all") {
+  stop(
+    "No method defined for get_draws for object of class(es) ",
+    paste(class(x), collapse = ", "),
+    ". Use directly on a wwinference_fit object or a",
+    "dataframe of wastewater observations.",
+    call. = FALSE
+  )
+}
+
+#' Vector of valid values for `what` in `get_draws`
+#' @noRd
+get_draws_what_ok <- c(
+  "all", "predicted_counts", "predicted_ww", "global_rt", "subpop_rt"
+)
+
+#' @rdname get_draws
+#' @param count_data A dataframe of the preprocessed daily count data (e.g.
+#' hospital admissions) from the "global" population
+#' @param stan_data_list A list containing all the data passed to stan for
+#' fitting the model
+#' @param fit_obj a CmdStan object that is the output of fitting the model to
+#' `x` and `count_data`
+#' @export
+get_draws.data.frame <- function(x,
+                                 count_data,
+                                 stan_data_list,
+                                 fit_obj,
+                                 ...,
+                                 what = "all") {
+  # Checking we are getting all
+  what_ok <- get_draws_what_ok
+
+  if (any(!what %in% what_ok)) {
+    idx <- which(!what %in% what_ok)
+    stop(
+      "The following invalid values were passed to `what`: ",
+      paste(what[idx], collapse = ", "), ". Valid values include: ",
+      paste(what_ok, collapse = ", "), "."
+    )
+  }
+
+  what_ok <- logical(length(what_ok))
+  names(what_ok) <- get_draws_what_ok
```

Comment:
I realized when working on #158 that if `stan_data_list$include_ww = 0` than we should not be allowing the user to get `subpop_rt` and `pred_ww`. The model as written currently will still return the prior, but I refactored it so that we are not carrying around an unnecessarily large model/latent variables for the hospital admissions only case, which means expected subpopulation R(t) estimates and predicted wastewater will not by default be generated.  Even if we do change this, I don't think we want "all" to return all 4 if its not fit to wastewater data. 

We probably need to add another layer of logic then that says something like:
```suggestion
  names(what_ok) <- ifelse(stan_data_list$include_ww ==0,  c("all", "predicted_counts", "global_rt"), get_draws_what_ok)
```

Reply:
Will open a separate issue for these two items!